### ISSUE TITLE
V1.3.4 There is a bug in the medusa speculate decoding ，Occasional bug

### ISSUE THREAD

**Issue by CSEEduanyu** [huggingface/text-generation-inference#1411]
### System Info

V1.3.4

### Information

- [X] Docker
- [X] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

like this 

### Expected behavior

This is an occasional bug and I can't reproduce it, but there are plenty of exceptions like this in production environments

**Comment by CSEEduanyu** [comment]
Often the first half sentence is normal, followed by an abnormal O or is. ", such strange characters

**Comment by OlivierDehaene** [comment]
@Narsil

**Comment by CSEEduanyu** [comment]
> @Narsil

@OlivierDehaene I'm pretty sure there is a bug when flashatten and mudusa are enabled at the same time. When batch=1, there will be no exception, but when the batch size is large, one of the tokens in the batch will appear strange

**Comment by CSEEduanyu** [comment]
![9429829fe7d36db9de5ffbcc34b03529382ce899](https://github.com/huggingface/text-generation-inference/assets/44152509/d520ef22-0706-4ee6-8824-326e3c41fce5)
When I turn off either flashatten or medusa, the exception will not appear.

**Comment by Narsil** [comment]
Could you share the model you used and something to reproduce ?

**Comment by CSEEduanyu** [comment]
In a translation scenario like this, the first few tokens will work, but when one token is abnormal, the result will be all wrong. And we find that the logits of the first token generated by the error are abnormal.

bad case1：
hello，today is OO，，，，，，，，，，，，，，，，，，，，，，，，，，。，，。，。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。000000"。"。"。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。 

bad case 2：
你好，今天天气不错奝OOOOOOOOOOOOO。O。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。。

**Comment by Narsil** [comment]
Could you upload your medusa 13B model somewhere ? so we can make sure we're talking about the same model.

**Comment by CSEEduanyu** [comment]
> 你能把你的 medusa 13B 模型上传到某个地方吗？所以我们可以确保我们谈论的是同一个模型。

Sorry, I can't upload my model because of company policy. The model is fine-tuned by llama2-13B on hugging face. (https://huggingface.co/meta-llama/Llama-2-13b-chat-hf). Our model was stable until the medusa version was updated, and I think you can reproduce this problem with Lema-2-13b-chat-HF. When we turn off flashatten, medusa will not be abnormal. If both are open at the same time, it will happen occasionally under heavy load (large batch).

**Comment by Narsil** [comment]
Ahh.. `--speculate` != medusa. Medusa are specially finetuned models with extra heads for speculation. Unless you produced one of those then it's not medusa but `ngram` speculation instead. (There are many ways to speculate).

**Comment by CSEEduanyu** [comment]
> Ahh.. `--speculate` != medusa. Medusa are specially finetuned models with extra heads for speculation. Unless you produced one of those then it's not medusa but `ngram` speculation instead. (There are many ways to speculate).
![image](https://github.com/huggingface/text-generation-inference/assets/44152509/bea167b8-245b-4674-89dc-20f90fe97e78)

I know, we put the trained medusa weights locally, I'm pretty sure we're using the medusa version. medusa our inference time consumption is reduced by about 50%.

**Comment by CSEEduanyu** [comment]
![image](https://github.com/huggingface/text-generation-inference/assets/44152509/f33fcba5-0ee9-47f9-b7a4-49fd8494808a)

**Comment by Narsil** [comment]
Ah ok so it's not just a fine-tuned it is a medusa fine-tune. Not having the real model under the hood makes it potential that I won't be able to reproduce the bug.

Do you know if you could reproduce with some integration tests we have (with your model).

```
pytest -sv --snapshot-update integration-tests/ -k test_flash_medusa
```

This is the test that is supposed to catch that kind of issue :https://github.com/huggingface/text-generation-inference/blob/main/integration-tests/models/test_flash_medusa.py#L52  (Althrough 4 out of 1000 might be rare enough it slipped through the cracks)

**Comment by CSEEduanyu** [comment]
> If you can reproduce with your model on that test setup at least that means I know I should be able to trigger it on my end with another dummy medusa model (ideally with the FasterDecoding already under test)
Ok, I'll try. I'll get back to you when we have the result. Thank you.

**Comment by CSEEduanyu** [comment]
> If you can reproduce with your model on that test setup at least that means I know I should be able to trigger it on my end with another dummy medusa model (ideally with the FasterDecoding already under test)

Ok, I'll try. I'll get back to you when we have the result. Thank you.
